### PR TITLE
add logic to find maximum allowed batch request size

### PR DESCRIPTION
**Pull request by DeveloperInProgress**
# Description
Add logic to adjust batch size for `JsonBatchRpcProvider`

Fixes https://github.com/subquery/subql/issues/1852

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have tested locally
- [ ] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [ ] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My code is up to date with the base branch
- [ ] I have updated relevant changelogs. [We suggest using chan](https://github.com/geut/chan/tree/main/packages/chan)

### CODE REVIEW

File: packages/node/src/ethereum/api.ethereum.ts [hunk 1]
```diff
@@ -137,6 +137,44 @@ export class EthereumApi implements ApiWrapper<EthereumBlockWrapper> {
...
+        batchSize = Math.floor((minBatchSize + maxBatchSize) / 2);
+      }
+
+      if (minBatchSize === maxBatchSize) {
+        found = true;
+      }
+    }
+
+    logger.debug(`batch request size for ${this.endpoint}: ${minBatchSize}`);
+    (this.client as JsonRpcBatchProvider).setBatchSize(minBatchSize);
```
Comment (stwiname) [inline packages/node/src/ethereum/api.ethereum.ts:177]: This doesn't seem like the right place to do this. It should live in the batch provider.

File: packages/node/src/ethereum/api.ethereum.ts [hunk 2]
```diff
@@ -137,6 +137,44 @@ export class EthereumApi implements ApiWrapper<EthereumBlockWrapper> {
...
+    let minBatchSize = 1;
+    let maxBatchSize = batchSize;
+    let found = false;
+
+    while (!found && minBatchSize <= maxBatchSize) {
+      try {
+        // Test the current batch size with a dummy request (getBlock)
+        const batchPromises = [];
+        for (let i = 0; i < batchSize; i++) {
+          batchPromises.push(this.client.getBlock('latest'));
```
Comment (stwiname) [inline packages/node/src/ethereum/api.ethereum.ts:160]: Ideally we don't waste requests doing this on startup each time and instead do it while making useful requests

**Review by stwiname — COMMENTED** [review]
We should add tests for this

File: packages/node/src/ethereum/ethers/json-rpc-batch-provider.ts [hunk 3]
```diff
@@ -112,8 +121,11 @@ export class JsonRpcBatchProvider extends JsonRpcProvider {
           batch.forEach((inflightRequest) => {
             inflightRequest.reject(error);
           });
+          this.failedBatchCount++;
+          this.adjustBatchSize();
```
Comment (stwiname) [inline packages/node/src/ethereum/ethers/json-rpc-batch-provider.ts:125]: These always are called in pairs (or with`this.successfulBatchCount++;`). We should combine them together to make sure that they are always called together

File: packages/node/src/ethereum/ethers/json-rpc-batch-provider.spec.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import { JsonRpcBatchProvider } from './json-rpc-batch-provider';
+import { ConnectionInfo } from './web';
+
+describe('JsonRpcBatchProvider', () => {
+  let batchProvider: JsonRpcBatchProvider;
+  let fetchJsonMock: jest.SpyInstance;
+
+  beforeEach(() => {
+    // Create a new instance of the JsonRpcBatchProvider before each test
+    batchProvider = new JsonRpcBatchProvider('http://localhost:8545');
```
Comment (stwiname) [inline packages/node/src/ethereum/ethers/json-rpc-batch-provider.spec.ts:11]: Is this endpoint non-existant to test failures?

**Review by stwiname — COMMENTED** [review]
This seems like a much better approach. The only problem I see is that we adjust the batch size for any error, not errors relating to ratelimits/batch sizes.

Its pretty common to make RPC requests to contract state that will throw an error, this doesn't mean that there was a problem with the batch size. Same goes if a block that doesn't exist is requested.

File: packages/node/src/ethereum/ethers/json-rpc-batch-provider.spec.ts (hunk 4 above)
Comment (DeveloperInProgress) [reply packages/node/src/ethereum/ethers/json-rpc-batch-provider.spec.ts:11]: we override the fetch method to resolve a promise locally, so no requests are made to the endpoint. it's just a placeholder

**Comment by DeveloperInProgress** [comment]
> The only problem I see is that we adjust the batch size for any error, not errors relating to ratelimits/batch sizes.

there does not seem to be a specific error for execeeding batch sizes. also, the error thrown on exceeding batch sizes seems to depend on the provider of the endpoint

**Comment by stwiname** [comment]
> there does not seem to be a specific error for execeeding batch sizes. also, the error thrown on exceeding batch sizes seems to depend on the provider of the endpoint

I think we can at least filter out expected errors then. Maybe we could test a few different RPC providers and cover most of the cases

File: packages/node/src/ethereum/api.ethereum.ts [hunk 5]
```diff
@@ -137,6 +137,13 @@ export class EthereumApi implements ApiWrapper<EthereumBlockWrapper> {
     this.supportsFinalization = supportsFinalization;
     this.chainId = network.chainId;
     this.name = network.name;
+
+    if (this.client instanceof WebSocketProvider) {
+      return;
+    }
+
+    // Find and set the maximum batch size allowed by JsonRpcBatchProvider
+    //await (this.client as JsonRpcBatchProvider).determineBatchSize();
```
Comment (stwiname) [inline packages/node/src/ethereum/api.ethereum.ts:146]: Can be reverted

File: packages/node/src/ethereum/ethers/json-rpc-batch-provider.ts [hunk 6]
```diff
@@ -146,6 +174,28 @@ export class JsonRpcBatchProvider extends JsonRpcProvider {
...
+  private adjustBatchSize(success: boolean) {
+    success ? this.successfulBatchCount++ : this.failedBatchCount++;
+    const totalBatches = this.successfulBatchCount + this.failedBatchCount;
+
+    if (totalBatches % this.batchSizeAdjustmentInterval === 0) {
+      const successRate = this.successfulBatchCount / totalBatches;
+
+      // Adjust the batch size based on the success rate.
+      if (successRate < 0.9 && this.batchSize > 1) {
+        this.batchSize--;
```
Comment (stwiname) [inline packages/node/src/ethereum/ethers/json-rpc-batch-provider.ts:191]: Do we have to limit the batch size to make sure it stays above 0

**Review by stwiname — COMMENTED** [review]
Changelog needs updating

File: packages/node/src/ethereum/ethers/json-rpc-batch-provider.ts (hunk 6 above)
Comment (DeveloperInProgress) [reply packages/node/src/ethereum/ethers/json-rpc-batch-provider.ts:191]: `if (successRate < 0.9 && this.batchSize > 1)` 

We have a check to make sure it stays above 0

**Review by stwiname — APPROVED** [review]

**Pull request merged**